numpyro/infer/hmc_gibbs.py

```diff
@@ -149,3 +153,178 @@ def potential_fn(z_gibbs, z_hmc):
         z = {**z_gibbs, **hmc_state.z}
 
         return HMCGibbsState(z, hmc_state, rng_key)
+
+
+def _discrete_step(rng_key, z_discrete, pe, potential_fn, idx, support_size):
+    # idx: current index of `z_discrete_flat` to update
+    # support_size: support size of z_discrete at the index idx
+    rng_key, rng_proposal, rng_accept = random.split(rng_key, 3)
+
+    # get z proposal
+    z_discrete_flat, unravel_fn = ravel_pytree(z_discrete)
+    proposal = random.randint(rng_proposal, (), minval=0, maxval=support_size - 1)
+    proposal = jnp.where(proposal >= z_discrete_flat[idx], proposal + 1, proposal)
+    z_discrete_new_flat = ops.index_update(z_discrete_flat, idx, proposal)
+    z_discrete_new = unravel_fn(z_discrete_new_flat)
+
+    pe_new = potential_fn(z_discrete_new)
+    accept_prob = jnp.clip(jnp.exp(pe - pe_new), a_max=1.0)
+    z_discrete, pe = cond(random.bernoulli(rng_accept, accept_prob),
+                          (z_discrete_new, pe_new), identity,
+                          (z_discrete, pe), identity)
+    return rng_key, z_discrete, pe
+
+
+def discrete_gibbs_fn(model, *model_args, **model_kwargs):
+    """
+    [EXPERIMENTAL INTERFACE]
+
+    Returns a gibbs_fn to be used in :class:`HMCGibbs`, which works for discrete latent sites
+    with enumerate support. Proposals are sampled uniformly over the set of possible values
+    in `enumerate_support` except for the current value (reference [1]).
+
+    Note that those discrete latent sites that are not specified in the constructor of
+    :class:`HMCGibbs` will be marginalized out by default (if they have enumerate supports).
+
+    **References**
+
+    1. *Peskun's theorem and a modified discrete-state Gibbs sampler*,
+       Liu, Jun S. (1996)
+
+    **Example**
+
+    .. doctest::
+
+        >>> from jax import random
+        >>> import jax.numpy as jnp
+        >>> import numpyro
+        >>> import numpyro.distributions as dist
+        >>> from numpyro.infer import MCMC, NUTS, HMCGibbs, discrete_gibbs_fn
+        ...
+        >>> def model(probs, locs):
+        ...     c = numpyro.sample("c", dist.Categorical(probs))
+        ...     numpyro.sample("x", dist.Normal(locs[c], 0.5))
+        ...
+        >>> probs = jnp.array([0.15, 0.3, 0.3, 0.25])
+        >>> locs = jnp.array([-2, 0, 2, 4])
+        >>> gibbs_fn = discrete_gibbs_fn(model, probs, locs)
```

Comment:
this interface seems a bit confusing since this isn't really a gibbs step. it's a mh step that apply its own mh correction; furthermore there's no complete conditional here.... maybe we need to rename some of these things?

Reply:
will switch to gibbs sampling here.

Reply:
@martinjankowiak I just updated the implementation of `_discrete_step` that uses a Gibbs update. Could you help me verify the math there?